docs/rules.md

```diff
@@ -0,0 +1,128 @@
+
+### 1. Function Hook Rule
+
+This is the most common rule type. It injects function calls at the beginning (`before`) and/or end (`after`) of a target function or method.
+
+**Use Cases:**
+
+- Wrapping functions with tracing spans.
+- Adding logging statements to function entries and exits.
+- Recording metrics about function calls.
+
+**Fields:**
+
+- `func` (string, required): The name of the target function to be instrumented.
+- `recv` (string, optional): The receiver type for a method. For a standalone function, this field should be omitted. For a pointer receiver, it should be prefixed with `*`, e.g., `*MyStruct`.
+- `before` (string, optional): The name of the function to be called at the entry of the target function.
+- `after` (string, optional): The name of the function to be called just before the target function returns.
+- `path` (string, required): The import path for the package containing the `before` and `after` hook functions.
+
+**Example:**
+
+```yaml
+hook_helloworld:
+  target: main
+  func: Example
+  before: MyHookBefore
+  after: MyHookAfter
+  path: "github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/helloworld"
+```
+
+This rule will inject `MyHookBefore` at the start of the `Example` function in the `main` package, and `MyHookAfter` at the end. The hook functions are located in the specified `path`.
+
+### 2. Struct Field Injection Rule
+
+This rule adds one or more new fields to a specified struct type.
+
+**Use Cases:**
+
+- Adding a context field to a struct to enable tracing through its methods.
+- Extending existing data structures with new information without modifying the original source code.
+
+**Fields:**
+
+- `struct` (string, required): The name of the target struct.
+- `new_field` (list of objects, required): A list of new fields to add to the struct. Each object in the list must contain:
+  - `name` (string, required): The name of the new field.
+  - `type` (string, required): The Go type of the new field.
+
+**Example:**
+
+```yaml
+add_new_field:
+  target: main
+  struct: MyStruct
+  new_field:
+    - name: NewField
+      type: string
+```
+
+This rule adds a new field named `NewField` of type `string` to the `MyStruct` struct in the `main` package.
+
+### 3. Raw Code Injection Rule
+
+This rule injects a string of raw Go code at the beginning of a target function. This offers great flexibility but should be used with caution as the injected code is not checked for correctness at definition time.
+
+**Use Cases:**
+
+- Injecting complex logic that cannot be expressed with a simple function call.
+- Quick and dirty debugging or logging.
+- Prototyping new instrumentation strategies.
+
+**Fields:**
+
+- `func` (string, required): The name of the target function.
+- `recv` (string, optional): The receiver type for a method.
+- `raw` (string, required): The raw Go code to be injected. The code will be inserted at the beginning of the target function.
```

Comment:
What about adding the raw Go code to the end of the target function?

Reply:
Semantically, `defer func(){ ....code }()` is equivalent to inserting code at the end of the function.